Split ATen/Parallel into interface and backend

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #20157 Support for Eigen thread pool
* #20087 Native ATen/Parallel backend
* **#20057 Split ATen/Parallel into interface and backend**
* #20050 Move inter-op settings into ATen/Parallel

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=OPENMP BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

Differential Revision: [D15248618](https://our.internmc.facebook.com/intern/diff/D15248618)

Commit:
Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

dzhulgakov commented on tools/build_pytorch_libs.py:
```diff
@@ -218,6 +218,11 @@ def run_cmake(version,
 
+    parallel_backend = os.getenv('PARALLEL_BACKEND')
+    if not parallel_backend:
+        parallel_backend = 'OPENMP'
```
don't you already have default in cmake? also, if it's not the - let's not add additional cmake_defines

dzhulgakov commented on aten/src/ATen/Parallel_common.cpp:
```diff
@@ -0,0 +1,56 @@
+#include <ATen/Parallel.h>
```
Maybe follow pattern and Parallel_common -> ParallelCommon and others?

dzhulgakov commented on aten/src/ATen/Parallel_openmp.h:
```diff
@@ -0,0 +1,108 @@
+#pragma once
+#include <ATen/ATen.h>
+#include <ATen/Parallel.h>
```
isn't it circular dep on Parallel.h?

ilia-cher replied:
right, no need to include this

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

cpuhrsch commented on aten/src/ATen/PTThreadPool.h:
```diff
@@ -0,0 +1,20 @@
+#pragma once
```
I wonder if at some point we want to move this into core.

ilia-cher replied:
shouldn't an issue to move this, as I understand..

ilia-cher replied:
oh, if you mean c10/core then c10 cannot depend on ATen, so that's why I had to move to ATen
soon it's not going to an issue as we're merging the libraries

cpuhrsch commented on aten/src/ATen/Parallel.h:
```diff
@@ -37,56 +26,17 @@ CAFFE2_API int get_num_threads();
-#endif
-}
+// Checks whether the code runs in parallel region
+CAFFE2_API bool in_parallel_region();
```
Is "CAFFE2_API" the right thing to use in aten?

cpuhrsch replied:
I guess it's used everywhere else...

ilia-cher replied:
It is, as of the moment ATen is compiled into libcaffe2

cpuhrsch commented on aten/src/ATen/Parallel.h:
```diff
@@ -122,43 +72,17 @@ two partial results and an identity.
 */
 template <class scalar_t, class F, class SF>
-inline scalar_t parallel_reduce(
+scalar_t parallel_reduce(
```
Are there any implications to not using inline?

ilia-cher replied:
I need to double check, possibly I can put it back, will check

cpuhrsch commented on aten/src/ATen/ParallelCommon.cpp:
```diff
@@ -0,0 +1,56 @@
+#include <mkl.h>
+#endif
+
+#ifdef _OPENMP
```
Does this use the build settings you created in an earlier PR to determine whether openmp is available or not? Maybe it makes sense to explicitly prevent us from including openmp, even if it is technically available as indicated by this macro.

ilia-cher replied:
I removed most of `_OPENMP` macro usage from PyTorch, except for a few places, including this - here it is just for the code below that prints information about libraries used, etc.

cpuhrsch replied:
Is there an advantage to / necessity for using _OPENMP?

ilia-cher replied:
I removed most of _OPENMP checks before because we expect the code to be agnostic of backend and just call ATen/Parallel function.

Here we include it because we call `omp_get_max_threads` explicitly below - only because I'd want to output OpenMP settings in `get_parallel_info`

ilia-cher replied:
So, no there's no advantage, it's the opposite - we'd want to not have them

ilia-cher replied:
@ezyang suggested adding a lint rule https://github.com/pytorch/pytorch/issues/20466

cpuhrsch commented on aten/src/ATen/ParallelCommon.cpp:
```diff
@@ -0,0 +1,56 @@
+  ss << "\tmkl_get_max_threads() : " << mkl_get_max_threads() << std::endl;
+#endif
+
+  ss << at::get_mkldnn_version() << std::endl;
```
What if mkldnn isn't available?

ilia-cher replied:
it will output "MKLDNN not found"

cpuhrsch commented on aten/src/ATen/ParallelOpenMP.cpp:
```diff
@@ -0,0 +1,80 @@
+  // Otherwise, MKL and our OpenMP-enabled functions will keep changing the
+  // size of the OpenMP thread pool, resulting in worse performance (and memory
+  // leaks in GCC 5.4)
+  omp_set_num_threads(mkl_get_max_threads());
```
What about the OMP_NUM_THREADS environment variable? A lot of users define it. Will that still work? What if MKL isn't available?

ilia-cher replied:
I did the analysis of how it works in https://github.com/pytorch/pytorch/issues/19001

cpuhrsch commented on aten/src/ATen/ParallelOpenMP.h:
```diff
@@ -0,0 +1,74 @@
+    int64_t chunk_size = divup((end - begin), num_threads);
+    int64_t begin_tid = begin + tid * chunk_size;
+    if (begin_tid < end) {
+      try {
```
Is a try catch block necessary here? Isn't it quite expensive?

ilia-cher replied:
could be expensive, in this PR I just refactored code into separate files though, to allow adding of other backends

cpuhrsch replied:
Ok, we should mark this for an immediate follow up PR, since it could slow down things a lot.

cpuhrsch commented on aten/src/ATen/ParallelOpenMP.h:
```diff
@@ -0,0 +1,74 @@
+      int64_t i = begin + id * grain_size;
+      results_data[id] = f(i, i + std::min(end - i, grain_size), ident);
+    }
+    return std::accumulate(
```
I've had bad experiences with the performance of std::accumulate. It might make sense to check that it's not terrible.

cpuhrsch replied:
If not resolved here, this should definitely be resolved in a follow-up PR or written down. I'd assume this will pop up in benchmarks eventuall.

ilia-cher replied:
replaced with a for loop

cpuhrsch commented on aten/src/ATen/ParallelThreadPoolNative.cpp:
```diff
@@ -0,0 +1,77 @@
+      "has started or set_num_interop_threads called");
+}
+
+int get_num_interop_threads() {
```
Why is this moving from the more general spot https://github.com/pytorch/pytorch/pull/20050/files#diff-c2f99b959224fab235eecd2cc4e696b8R160 to here?

ilia-cher replied:
because the implementation of `get_num_interop_threads` depends on how we implement inter-op parallelism - we have two choices - native thread pool (here) and (potentially) TBB

ilia-cher replied:
so ParallelThreadPoolNative.cpp specifically implements inter-op parallelism using a native thread pool

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

cpuhrsch commented on aten/src/ATen/ParallelOpenMP.h:
```diff
@@ -0,0 +1,63 @@
+    const int64_t num_results = divup((end - begin), grain_size);
+    std::vector<scalar_t> results(num_results);
+    scalar_t* results_data = results.data();
+#pragma omp parallel for if ((end - begin) >= grain_size)
```
Do we not want to guard this using _OPENMP as well?

ilia-cher replied:
we used `_OPENMP` above because of `omp_get_num_threads`, `omp_get_thread_num` from `omp.h` which include is guarded by `_OPENMP`

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

Commit:
Update on "Split ATen/Parallel into interface and backend"

Split ATen/Parallel into interface and backend

Summary:
Split into ATen/Parallel into backend-independent part and
specific implementations (currently - OpenMP and native thread pool).
Also adding cmake option to choose corr. backend

Test Plan:
PARALLEL_BACKEND=ABC BLAS=MKL USE_MKLDNN=1 USE_OPENCV=1 USE_FFMPEG=1
python setup.py develop --cmake

gh-metadata: pytorch pytorch 20057 gh/ilia-cher/14/head

ezyang commented:
I believe this leads to SIGFPE in 

```
May 22 16:00:16 test_empty_dropout_cuda (__main__.TestNN) ... Traceback (most recent call last):
May 22 16:00:16   File "test/run_test.py", line 440, in <module>
May 22 16:00:16     main()
May 22 16:00:16   File "test/run_test.py", line 432, in main
May 22 16:00:16     raise RuntimeError(message)
May 22 16:00:16 RuntimeError: test_nn failed! Received signal: SIGFPE
```

ilia-cher commented:
@ezyang let me double check, haven't noticed that before

suo commented:
Also I think it breaks test_cpp_extensions:
```
May 22 07:21:20 cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++
May 22 07:21:20 cuda_extension.cpp: In function 'at::Tensor sigmoid_add(at::Tensor, at::Tensor)':
May 22 07:21:20 cuda_extension.cpp:9:60: error: 'TORCH_CHECK' was not declared in this scope
May 22 07:21:20    TORCH_CHECK(x.type().is_cuda(), "x must be a CUDA tensor");
May 22 07:21:20                                                             ^
May 22 07:21:20 error: command 'gcc' failed with exit status 1
May 22 07:21:20 Traceback (most recent call last):
May 22 07:21:20   File "test/run_test.py", line 440, in <module>
May 22 07:21:20     main()
May 22 07:21:20   File "test/run_test.py", line 432, in main
May 22 07:21:20     raise RuntimeError(message)
May 22 07:21:20 RuntimeError: test_cpp_extensions failed!
```

Can we unland?

ezyang commented:
It was unlanded

suo commented:
oh, github CI hooks are broken. Was wondering why I couldn't see it on the pytorch hud
________________________________
From: Edward Z. Yang <notifications@github.com>
Sent: Wednesday, May 22, 2019 11:15 AM
To: pytorch/pytorch
Cc: Michael Suo; Comment
Subject: Re: [pytorch/pytorch] Split ATen/Parallel into interface and backend (#20057)


It was unlanded

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/pytorch/pytorch/pull/20057?email_source=notifications&email_token=AAMK4EDOKHJSZ4Q3EZ67SRLPWWEV3A5CNFSM4HJ4SZM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV73PKY#issuecomment-494909355>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAMK4EA4QIQ3TECYS3MIU7TPWWEV3ANCNFSM4HJ4SZMQ>.

ilia-cher commented:
(resending as https://github.com/pytorch/pytorch/pull/20825)
